### PR TITLE
Add logics for deleting tasks and marking tasks as done/undone

### LINKED ISSUE #13: As a user, I want to see when a task was done
... so that I can evaluate my improvement over a certain period of time.

_User story for epic in issue #7 _

### PR DESCRIPTION
**Pull request by jiayushe**
- Allow adding of tasks
- Allow deleting of tasks
- Allow marking tasks as done/undone
==========
- Fix #13 #51 #52 #53
==========
- JSON storage for tasks and plans in next PR :)

### CODE REVIEW

File: src/main/java/seedu/algobase/logic/commands/DeleteTaskCommand.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_PLAN + "1 "
+            + PREFIX_TASK + "10";
+
+    public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
+
+    private final DeleteTaskDescriptor deleteTaskDescriptor;
+
+    /**
+     * Creates a DeleteTaskCommand to delete a {@code Task} to the specified {@code Plan}
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/DeleteTaskCommand.java]: typo: "in" the specified Plan.

File: src/main/java/seedu/algobase/logic/commands/DeleteTaskCommand.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        if (deleteTaskDescriptor.planIndex.getZeroBased() >= lastShownPlanList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+        }
+
+        Plan planToUpdate = lastShownPlanList.get(deleteTaskDescriptor.planIndex.getZeroBased());
+        Set<Task> taskSet = planToUpdate.getTasks();
+        List<Task> taskList = new ArrayList<>(taskSet);
+        Task task = taskList.get(deleteTaskDescriptor.taskIndex.getZeroBased());
+        taskList.remove(deleteTaskDescriptor.taskIndex.getZeroBased());
+        taskSet = new HashSet<>(taskList);
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/DeleteTaskCommand.java]: Can be refactored to:
```java
List<Task> taskList = new ArrayList<>(planToUpdate.getTasks());
// ... omitted
Set<Task> taskSet = new HashSet<>(taskList);
```
to make the logic look clearer.

File: src/main/java/seedu/algobase/logic/commands/DoneTaskCommand.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if (doneTaskDescriptor.planIndex.getZeroBased() >= lastShownPlanList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+        }
+
+        Plan planToUpdate = lastShownPlanList.get(doneTaskDescriptor.planIndex.getZeroBased());
+        Set<Task> taskSet = planToUpdate.getTasks();
+        List<Task> taskList = new ArrayList<>(taskSet);
+        Task task = taskList.get(doneTaskDescriptor.taskIndex.getZeroBased());
+        taskList.remove(doneTaskDescriptor.taskIndex.getZeroBased());
+        taskSet = new HashSet<>(taskList);
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/DoneTaskCommand.java]: Similar to last comment.

File: src/main/java/seedu/algobase/logic/commands/UndoneTaskCommand.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        if (undoneTaskDescriptor.planIndex.getZeroBased() >= lastShownPlanList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+        }
+
+        Plan planToUpdate = lastShownPlanList.get(undoneTaskDescriptor.planIndex.getZeroBased());
+        Set<Task> taskSet = planToUpdate.getTasks();
+        List<Task> taskList = new ArrayList<>(taskSet);
+        Task task = taskList.get(undoneTaskDescriptor.taskIndex.getZeroBased());
+        taskList.remove(undoneTaskDescriptor.taskIndex.getZeroBased());
+        taskSet = new HashSet<>(taskList);
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/UndoneTaskCommand.java]: Needs reordering.

File: src/main/java/seedu/algobase/logic/parser/AlgoBaseParser.java [hunk 5]
```diff
@@ -56,53 +59,59 @@ public Command parseCommand(String userInput) throws ParseException {
...
 
-        case FindPlanCommand.COMMAND_WORD:
-            return new FindPlanCommandParser().parse(arguments);
+        case SwitchCommand.COMMAND_WORD:
+            return new SwitchCommandParser().parse(arguments);
 
-        case ListPlanCommand.COMMAND_WORD:
-            return new ListPlanCommand();
+        case UndoneTaskCommand.COMMAND_WORD:
+            return new UndoneTaskCommandParser().parse(arguments);
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/parser/AlgoBaseParser.java]: These commands can be reorganized like @tiuweehan did.

File: src/main/java/seedu/algobase/logic/parser/CliSyntax.java [hunk 6]
```diff
@@ -16,8 +16,9 @@
     public static final Prefix PREFIX_SOURCE = new Prefix("src/");
     public static final Prefix PREFIX_SORTING_METHOD = new Prefix("m/");
     public static final Prefix PREFIX_SORTING_ORDER = new Prefix("ord/");
-    public static final Prefix PREFIX_PLAN = new Prefix("pl/");
-    public static final Prefix PREFIX_PROBLEM = new Prefix("pr/");
+    public static final Prefix PREFIX_PLAN = new Prefix("plan/");
+    public static final Prefix PREFIX_PROBLEM = new Prefix("pro/");
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/parser/CliSyntax.java]: Probably `pro` is misleading? Can we use `prob`?

**Review by le0tan — CHANGES_REQUESTED** [review]

**Review by tiuweehan — APPROVED** [review]

**Review by le0tan — APPROVED** [review]
lgtm

**Pull request merged**